gateway-server/src/main/java/org/apache/knox/gateway/services/security/impl/ZookeeperRemoteAliasService.java

```diff
@@ -169,19 +171,23 @@ private static void checkPathsExist(
   public List<String> getAliasesForCluster(final String clusterName) throws AliasServiceException {
     List<String> remoteAliases = new ArrayList<>();
 
-    /* If we have remote registry configured, query it */
-    if (remoteClient != null) {
-      remoteAliases = remoteClient
-          .listChildEntries(buildClusterEntryName(clusterName));
+    // First, check the alias cache
+    if (aliasCache.containsKey(clusterName)) {
+      remoteAliases.addAll(aliasCache.get(clusterName).keySet());
+    } else { // If it's not available locally, then check ZK
+      /* If we have remote registry configured, query it */
+      if (remoteClient != null) {
+        remoteAliases = remoteClient.listChildEntries(buildClusterEntryName(clusterName));
+      }
```

Comment:
I don't think I follow this logic. In my head, the cache should be up to date. We should never just straight return remove data without updating the cache.

Reply:
If the cache isn't up to date for whatever reason, we should update the cache and return that data. Now have to be careful about a race condition here.